Track grades for Blockstore content, emit tracking logs

bradenmacdonald opened the pull request:
Follow up to https://github.com/edx/edx-platform/pull/21630 . This one allows Blockstore-hosted XBlocks to emit grades and tracking log events. Completion tracking will come in a separate PR since it builds on this but also requires changes to an external package as well as a DB migration.

The goal of this is to store scores for individual blocks in CSM, but not to impact the broader persistent grades / subsection grades feature which should just ignore these non-course scores.

### Test instructions

Manual test:
* Run Blockstore (`cd blockstore; make easyserver`)
* Install the very latest commit of [Ramshackle](https://github.com/open-craft/ramshackle/)
* Log in to Studio and the LMS
* Go to http://localhost:18010/ramshackle/ and create a content library with a capa problem
* Submit an answer
* Find that problem's entry in http://localhost:18000/admin/courseware/studentmodule/ and verify that its "grade" and "max grade" are updated (without this PR they will be blank)
* Verify that the LMS log output a tracking event like this, with `context.context_id` set and `context.course_id` blank:
    ```
    2019-09-17 22:48:55,462 INFO 31833 [tracking] [user None] logger.py:50 - 
       {
    "username": "",
    "event_type": "problem_check",
    "ip": "172.18.0.1",
    "agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36",
    "host": "localhost:18000",
    "referer": "http://d3749cj02gkez2.cloudfront.net/xblock-bootstrap.html",
    "accept_language": "en;q=1.0, en;q=0.9",
    "event": {
        "submission": {...},
        "success": "correct",
        "grade": 1,
        "correct_map": {...},
        "state": {...},
        "answers": {
        "lb:LX:test-creation2:problem:problem1_2_1": "choice_1"
        },
        "attempts": 4,
        "max_grade": 1,
        "problem_id": "lb:LX:test-creation2:problem:problem1"
    },
    "event_source": "server",
    "context": {
        "user_id": 9,
        "org_id": "LX",
        "asides": {},
        "context_id": "lib:LX:test-creation2",
        "course_id": "",
        "path": "/api/xblock/v2/xblocks/lb:LX:test-creation2:problem:problem1/handler/9-7453c2d14d7d0cf90ca8/xmodule_handler/problem_check"
    },
    "time": "2019-09-17T22:48:55.462504+00:00",
    "page": "x_module"
    }

    ```

Management command impact:
* Verify that the `recalculate_subsection_grades` task runs correctly by running `./manage.py lms recalculate_subsection_grades --modified_start '2019-01-01 16:43' --modified_end '2019-12-25 16:43'`

Unit tests:
* Run `make testserver` from the `blockstore` dir, then from both `lms-shell` and `studio-shell`, run:
    ```
    EDXAPP_RUN_BLOCKSTORE_TESTS=1 python -Wd -m pytest --ds=lms.envs.test openedx/core/lib/blockstore_api/ openedx/core/djangolib/tests/test_blockstore_cache.py openedx/core/djangoapps/content_libraries/tests/ common/lib/xmodule/xmodule/tests/test_unit_block.py
    ```
(Change `lms.envs.test` to `cms.envs.test` when running in `studio-shell`)

ormsbee commented on openedx/core/djangoapps/xblock/runtime/runtime.py:
```diff
@@ -94,8 +111,74 @@ def local_resource_url(self, block, uri):
+        else:
+            # Why don't we also log events even when special_handler is called?
+            # Not sure, but this if/else is what the similar code in module_render.py was doing.
+            self.log_event_to_tracking_log(block, event_type, event_data)
```
I thought the idea was to allow specific events to fill in their own details while still having a generic fallback?

bradenmacdonald replied:
Yeah, I just found it weird because in this file it doesn't look like any log gets emitted for grades, but if you follow the logic through the various signals and handlers, it does end up emitting the log. I've removed this comment while squashing.

bradenmacdonald commented:
Thanks @ormsbee! Squashed and rebased in preparation for merge.

> So just sanity checking, but we're just adding the context_id key for events here, and not making any backwards incompatible changes to events, right?

Yes, in fact for maximum backwards compatibility I didn't even add the `context_id` to to course events - only events from blockstore libraries will have that additional field for now. Events in the existing runtime should be completely unchanged.

edx-webhook commented:
Although this pull request is already merged, I've created [OSPR-5410](https://openedx.atlassian.net/browse/OSPR-5410) so that we can track it in Jira.

There is nothing you have to do. No action is needed from your side.  Thanks again for your contribution.

<!-- comment:welcome_closed -->
<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->
